Parser/pegen/pegen.c

```diff
@@ -490,90 +490,96 @@ number_token(Parser *p)
                     p->arena);
 }
 
-PyObject *
-run_parser(struct tok_state *tok, void *(start_rule_func)(Parser *), int mode, int input_mode,
-           KeywordToken **keywords, int n_keyword_lists)
+void
+Parser_Free(Parser *p)
+{
+    for (int i = 0; i < p->size; i++) {
+        PyMem_Free(p->tokens[i]);
+    }
+    PyMem_Free(p->tokens);
+    PyMem_Free(p);
+}
+
+Parser *
+Parser_New(struct tok_state *tok, mod_ty(*parse_func)(Parser *), int input_mode,
+           PyArena *arena)
 {
-    PyObject *result = NULL;
     Parser *p = PyMem_Malloc(sizeof(Parser));
     if (p == NULL) {
         PyErr_Format(PyExc_MemoryError, "Out of memory for Parser");
-        goto exit;
+        return NULL;
     }
     assert(tok != NULL);
     p->tok = tok;
     p->input_mode = input_mode;
-    p->keywords = keywords;
+    p->keywords = reserved_keywords;
     p->n_keyword_lists = n_keyword_lists;
     p->tokens = PyMem_Malloc(sizeof(Token *));
     if (!p->tokens) {
         PyErr_Format(PyExc_MemoryError, "Out of memory for tokens");
-        goto exit;
+        return NULL;
     }
     p->tokens[0] = PyMem_Malloc(sizeof(Token));
     memset(p->tokens[0], '\0', sizeof(Token));
     p->mark = 0;
     p->fill = 0;
     p->size = 1;
 
-    p->arena = PyArena_New();
-    if (!p->arena) {
-        goto exit;
+    if (arena == NULL) {
+        p->arena = PyArena_New();
+        if (!p->arena) {
+            return NULL;
+        }
+    } else {
+        p->arena = arena;
     }
 
     if (fill_token(p) < 0) {
-        goto exit;
+        return NULL;
     }
 
-    PyErr_Clear();
+    p->start_rule_func = parse_func;
 
-    p->start_rule_func = start_rule_func;
+    return p;
+}
 
+void *
+run_parser(Parser *p, int mode)
```

Comment:
Probably we should just completely get rid of `mode`, and just always return a `mod_ty`. The modes other than 1 were used to measure performance better, but I think we'll soon have a different way for that.

Reply:
That was what I had worked out prior to [`15bb3a2`](https://github.com/we-like-parsers/cpython/pull/11/commits/a5bb3a2aefe43ba581e4ada6f7f06494e1b930d7). All of the `run_parser*` functions were returning a `mod_ty` and the conversion to a code object or an AST object was in the C API functions themselves. This was segfaulting though due to `PyAST_CompileObject` expecting the filename as well. An obvious solution to this is to setup the filename object in the C API functions, but that seems a bit ugly to me. Any other ideas?

Reply:
We do something similar now, the `filename` is always passed down even if you are using `<string>`:

https://github.com/python/cpython/blob/master/Python/pythonrun.c#L1025-L1033

Reply:
> were returning a `mod_ty` and the conversion to a code object or an AST object was in the C API functions themselves.

In general, even if it would look more ugly (maybe it does not with the proper design) having a function return different things depending on the input is also dangerous, because you lose a lot of the type safety (remember all the debugging we needed to do for the grammar rules because we didn't have type safety and we were returning incorrect things).

Reply:
> We do something similar now, the `filename` is always passed down even if you are using `<string>`:
> 
> https://github.com/python/cpython/blob/master/Python/pythonrun.c#L1025-L1033

Both of your points are true, yeah. I'll begin working on refactoring the `run_parser*` functions in a bit. Thank you both for your feedback on this.